value_test.go

```diff
@@ -757,16 +755,64 @@ func TestPenultimateLogCorruption(t *testing.T) {
 	db1, err := Open(opt)
 	require.NoError(t, err)
 	h.db = db1
-	h.readRange(0, 1) // Only 2 should be gone, because it is at the end of logfile 0.
-	h.readRange(3, 7)
+	h.readRange(0, 6) // Only 7 should be gone, because it is at the end of logfile 2.
 	err = db1.View(func(txn *Txn) error {
-		_, err := txn.Get(h.key(2)) // Verify that 2 is gone.
+		_, err := txn.Get(h.key(7)) // Verify that 7 is gone.
 		require.Equal(t, ErrKeyNotFound, err)
 		return nil
 	})
 	require.NoError(t, err)
 	require.NoError(t, db1.Close())
 }
+func TestForceFlushMemtable(t *testing.T) {
+	// This test checks, if latest value of value log head is getting persisted
+	// after value log file rotation. Latestt value is persisted by flushing current
+	// memtable to disk.
+
+	dir, err := ioutil.TempDir("", "badger")
+	require.NoError(t, err)
+	defer os.RemoveAll(dir)
+	opt := getTestOptions(dir)
+	opt.ValueLogLoadingMode = options.FileIO
+	opt.ValueLogMaxEntries = 1
+
+	db0, err := Open(opt)
+	require.NoError(t, err)
+
+	// write two entries to db. Since there can be max 1 entry in
+	// each value log file, there will be two log file rotation. latest value
+	// persisted for head will have file id as 1.
+	for i := 0; i < 2; i++ {
+		err := db0.Update(func(txn *Txn) error {
+			return txn.Set([]byte(fmt.Sprintf("%d", i)), []byte(fmt.Sprintf("%d", i)))
+		})
+		require.NoError(t, err)
+	}
+
+	// We want to check persisted value of value log head, hence open another instance of
+	// db from same opt. Release directory lock of first instance
+	if db0.dirLockGuard != nil {
+		require.NoError(t, db0.dirLockGuard.release())
+	}
+	if db0.valueDirGuard != nil {
+		require.NoError(t, db0.valueDirGuard.release())
+	}
+
+	// wait for memtable to flush to disk
```

Comment:
Is this a deterministic test?

Reply:
It is passing every time on my local system. Lets see how it behaves on teamcity. Otherwise can increase wait time.

Reply:
Just mention it here in that case on the test so that in case it ever fails, somebody else can fix it too.